docs/DesignDocs/openenclave_modules.md

```diff
@@ -0,0 +1,105 @@
+# OpenEnclave Modules
+
+The Open Enclave SDK provides a core set of functionality for build enclave
+applications. Some of these features&2014;such as memory management and ecall/ocall
+marshalling&2014;are required to create an enclave. Other features such as
+libc functions, host filesystem, and network sockets can be optionally included
+inside the enclave by the enclave developer. These optional features are known
+as **Open Enclave Modules** and are static libraries that can be optionally
+linked into the enclave binary.
+
+## Currently supported modules
+
+Currently, Open Enclave only supports modules that are part of the OE SDK source tree.
+Today there are 5 optionally includable modules:
+
+* hostfs
+* libc
+* hostepoll
+* hostsock
+* hostresolver
+
+## Module initialization
+
+OE modules can each have an initialization function which will be called on
+first entry to the enclave. This ensures that a module are fully initialized
+prior to an application using it. There are two ways to demark initialization
+functions:
+
+1. `OE_MOD_INIT Attribute`
+
+The `OE_MOD_INIT` attribute can be applied to a function. This ensures that the
+function is called on enclave entry.
+
+```
+void OE_MOD_INIT init_my_module(void)
+{
+    // Do stuff
+}
+```
+
+2. `OE_MOD_INIT_PRI` Attribute
+
+Optionally, modules can specify a priority for their initialization functions.
+Functions with a lower priority value will be called before functions with a
+higher priority value. This can be used if one module depends on another being
+initialized first.
+
+```
+// mod1.c
+void OE_MOD_INIT_PRI(0) mod1_init(void)
+{
+    // Initialize module 1
+}
+
+// mod2.c
+void OE_MOD_INIT_PRI(1) mod2_init(void)
+{
+    // Do stuff that depends on mod1. This function will
+    // always be called after mod1_init.
+}
+```
+
+### Linker disclaimer
+
+Because modules can be added and removed at compile-time without the core OE
+library knowing about them, initialization functions are not actually
+referenced by symbol. They are instead kept in their own section in the binary.
+To ensure a module's initialization function is included in the final enclave
+binary there are 3 options:
+
+1. Add the module initialization function in the same file as a symbol that is
+referenced in the enclave binary.
+
+```
+// mod_init.c
+
+int always_used_symbol = 1;
+
+void OE_MOD_INIT mod_init(void)
+{
+    // Do initialization
+}
+
+// enclave_app.c
+int main_function()
+{
+    int val = always_used_symbol;
+    // Do stuff
+}
+```
+
+2. Pass a flag to tell the linker to always link the initialization function.
+
+```
+void OE_MOD_INIT mod_init(void)
+{
+    // Do initialization
+}
+```
+
+Then during compilation `$CC -o enclave -lmymodule.a -W,l,-u mod_init`
```

Comment:
--require-defined would be a better option than -u since the former would error out if the symbol is not found whereas the latter won't.